UnmanagedCallersOnlyAttribute Support

joncham opened the pull request:
<!--
Thank you for your Pull Request!

Here are a few things to think about (see below for more details). Please check each option after the PR is created.
-->

- Should this pull request have release notes?
  - [X] Yes
  - [ ] No
- Do these changes need to be back ported?
  - [ ] Yes
  - [X] No
- Do these changes need to be upstreamed to [mono/mono](https://github.com/mono/mono) or [dotnet/runtime](https://github.com/dotnet/runtime) repositories?
  - [ ] Yes
  - [X] No

Reviewers: please consider these questions as well! :heart:

**Release notes**

Internal @joncham:
Mono: Add UnmanagedCallersOnlyAttribute support

<!-- Most pull requests should have release notes.

Use Internal for release notes that should not be public.

Other options: Changed, Improved, Feature.
-->

<!-- Use this section is the pull request should be back ported.
**Backports**

List the versions of Unity where this change should be back ported here.
-->

<!-- Use this section if the pull request requires other changes in the Unity repository.
**Unity repository changes**

List any Unity repository PRs.
-->

joncham committed:
Detect UnmanagedCallersOnlyAttribute by name.

This enables the codepath in !ENABLE_NETCORE where the user defines the
attribute with matching name and namespace.

joncham committed:
Add calling convention support.

Inspired by https://github.com/dotnet/runtime/commit/549b4314c9ea92c124b5a8dc88a881f7fdc52274

joncham commented on mono/metadata/marshal.c:
```diff
@@ -6518,7 +6603,34 @@ gboolean
+	ERROR_DECL (attr_error);
+	MonoClass* attr_klass = NULL;
+	/* attribute is only allowed on static methods */
+	if (!m_method_is_static(method))
```
Should we keep these early outs to speed up this check? Or should we let everything through and allow later "error" checks to handle these cases?

UnityAlex replied:
I'm fine with keeping the early outs.

scott-ferguson-unity replied:
Roslyn will error on these cases too, so only users doing bad custom IL should hit these failure cases, so as long as we do something other than randomly crash in hard to debug ways we should be fine.  What is the failure mode in those cases?

bholmes replied:
Should the early outs be in mono_marshal_set_callconv_from_unmanaged_callers_only_attribute as well. 

I feel that mono_method_has_unmanaged_callers_only_attribute and mono_marshal_set_callconv_from_unmanaged_callers_only_attribute duplicate code and are not consistent?

joncham replied:
Ideally, I would not have early outs. I'm just a bit worried about the cost of the attribute checks. Let me evaluate any impact again, and try to just leave a regular code path.

joncham replied:
I have removed the early outs, this means we will hit error code paths at other places in mono code. For example, when the ldftn is encountered it validates proper use of the attribute.

If we have specific error cases we care about, we should write tests in IL2CPP or Unity, and ensure all runtimes properly handle them.

UnityAlex commented:
Judging by the NETCORE checks I am guessing some of this is used upstream already? Just trying to get a feel for if this is used in production already or not.

joncham commented:
> Judging by the NETCORE checks I am guessing some of this is used upstream already? Just trying to get a feel for if this is used in production already or not.

The first commit I made myself. The second commit was inspired by upstream but quite modified to avoid taking a bunch of changes.

UnityAlex commented on mono/metadata/marshal.c:
```diff
@@ -3264,6 +3265,88 @@ mono_marshal_set_callconv_from_modopt (MonoMethod *method, MonoMethodSignature *
+	ERROR_DECL(error);
+	MonoCustomAttrInfo* cinfo = mono_custom_attrs_from_method_checked (method, error);
+	if (!is_ok(error) || !cinfo) {
+		mono_error_cleanup(error);
```
Doesn't look like `mono_error_cleanup` is called if we get past here.

joncham replied:
I think that only needs called when an error happens. I have added checks after each function taking an error, and just `mono_error_assert_ok` for now. I'd rather crash and detect we are doing something wrong I think.

scott-ferguson-unity commented on mono/metadata/marshal.c:
```diff
@@ -3264,6 +3265,88 @@ mono_marshal_set_callconv_from_modopt (MonoMethod *method, MonoMethodSignature *
+	else if (!strcmp (name, "SuppressGCTransition"))
+		sig->suppress_gc_transition = 1;
+	// TODO: Support CallConvMemberFunction?
+	// TODO: Support multiple calling convetions?
```
I played multiple calling conventions in CoreCLR and it only supports multiple calling conventions when they don't conflict.  e.g. CallConvStdCall + CallConvMemberFunction is OK, but CallConvStdCall + CallConvCdecl throws:
> System.InvalidProgramException: 'Multiple unmanaged calling conventions are specified. Only a single calling convention is supported.'

scott-ferguson-unity commented:
Also `UnmanagedCallersOnlyAttribute` isn't defined in our classlibs - we'll have to add it too.

joncham commented:
> Also `UnmanagedCallersOnlyAttribute` isn't defined in our classlibs - we'll have to add it too.

Why do we need to add it?

mrvoorhe reviewed (changes requested):
Changes are needed to make this work with all managed code stripping levels.

coreclr ensures that `UnmanagedCallersOnlyAttribute` survives using the embedded linker descriptor file in `System.Private.CoreLib` https://github.com/Unity-Technologies/runtime/blob/1cd7eb4816b773f35cc75e522e4339525a2a8c51/src/mono/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.xml#L635

We don't have that logic.

And if we don't add the attribute to our class libraries, then we wouldn't be able to mirror the approach used by coreclr.  Instead we'd have to add logic to UnityLinker.

mrvoorhe reviewed (changes requested):
Changes are needed to make this work with all managed code stripping levels.

coreclr ensures that `UnmanagedCallersOnlyAttribute` survives using the embedded linker descriptor file in `System.Private.CoreLib` https://github.com/Unity-Technologies/runtime/blob/1cd7eb4816b773f35cc75e522e4339525a2a8c51/src/mono/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.xml#L635

We don't have that in our [mscorlib.xml](https://github.com/Unity-Technologies/mono/blob/unity-main/mcs/class/corlib/LinkerDescriptor/mscorlib.xml)

And if we don't add the attribute to our class libraries, then we wouldn't be able to mirror the approach used by coreclr.  Instead we'd have to add logic to UnityLinker.

joncham commented:
@mrvoorhe I'm not sure what the requested change is? We need to support this attribute defined by anyone in any assembly, similar to `MonoPInvokeCallbackAttribute`.

mrvoorhe commented:
> @mrvoorhe I'm not sure what the requested change is? We need to support this attribute defined by anyone in any assembly, similar to `MonoPInvokeCallbackAttribute`.

If you want to go that route that is fine.  In that case no changes would be needed to this PR, but a PR to UnityLinker should be made to add the same logic as `MonoPInvokeCallbackAttribute`


https://github.cds.internal.unity3d.com/unity/il2cpp/blob/main/UnityLinker/Linker.Steps/Marking/CommonRuntimeMarking.cs
https://github.cds.internal.unity3d.com/unity/il2cpp/blob/main/Unity.Linker.Tests.Cases/Il2CppRuntime/Attributes/OnlyKeepUsed/MonoPInvokeCallbackAttributeIsKept.cs
https://github.cds.internal.unity3d.com/unity/il2cpp/blob/main/Unity.Linker.Tests.Cases/MonoRuntime/Attributes/OnlyKeepUsed/MonoPInvokeCallbackAttributeIsKept.cs

scott-ferguson-unity commented:
> > Also `UnmanagedCallersOnlyAttribute` isn't defined in our classlibs - we'll have to add it too.
> 
> Why do we need to add it?

Wouldn't we need to to compile/resolve any of the Mono/IL2CPP usages of the attribute we're planning to add?  Or can we leave it out as long as we're doing IL injection?

bholmes reviewed (approved):
If the testing looks good, ship it!

joncham commented:
> Wouldn't we need to to compile/resolve any of the Mono/IL2CPP usages of the attribute we're planning to add? Or can we leave it out as long as we're doing IL injection?

We can add it if the class libraries themselves need it. But for Unity usage, we already define it in UnityEngine for when we target CoreCLR. We will match on name and not assembly (like we do other attributes).

Once we target .NET as a compilation target (instead of .NET Standard and .NET Framework+) we can use the one provided by the profile itself.

joncham committed:
Update RuntimeMethodHandle.GetFunctionPointer to support UnmanagedCallersOnly.

Adapted from https://github.com/dotnet/runtime/commit/bda872709d797a0149349a8d4531774921e7a99b
